messagequeue/chunker.go

```diff
@@ -74,32 +43,36 @@ type Handler[T any] func(chunk []T)
 // slices of a specified size from the passed buffer and pass them to the handler.
 // Entries retrieved from the buffer are guaranteed to be delivered to
 // the handler within the configured ChunkElementExpiry duration.
-func NewChunker[T any](buffer <-chan T, handler Handler[T],
-	options *ChunkerOptions) Chunker[T] {
+func NewChunker[T any](options *ChunkerOptions) Chunker[T] {
 	if nil == options {
 		options = NewChunkerOptions()
 	}
 	return &chunker[T]{
 		options: options,
-		buffer:  buffer,
-		// we need a control channel since we are not controlling the
-		// buffer
-		control: make(chan bool),
-		handler: handler,
 		state:   &atomic.Uint32{},
 	}
 }
 
-func (c *chunker[T]) Start() error {
+func (c *chunker[T]) Start(buffer <-chan T, handler Handler[T]) error {
+	if nil == buffer {
+		return errors.New("buffer cannot be nil")
+	}
+	if nil == handler {
+		return errors.New("handler cannot be nil")
+	}
 	c.mux.Lock()
 	defer c.mux.Unlock()
 	if !c.state.CompareAndSwap(stateStopped, stateRunning) {
 		return errors.New("Chunker.Run called while not in state 'Stopped'")
 	}
+	c.buffer = buffer
+	c.handler = handler
 	// validate our options
-	if err := c.options.Valid(); err != nil {
+	if err := c.options.Validate(); err != nil {
 		return err
 	}
+	c.buffer = buffer
+	c.handler = handler
```

Comment:
Is there a reason you are setting buffer and handler again?

Reply:
refactor artifact. good catch :+1: